### PR TITLE
Add enough feature to be able to parse clojure.core

### PR DESCRIPTION
**Pull request by plexus**
- metadata ^
- deref @...
- shorthand lambdas #(..)
- vars #'
- hex numbers
- regexes

### CODE REVIEW

File: test/parseclj-test.el [hunk 1]
```diff
@@ -185,6 +185,73 @@
...
+                         (:position . 1)
+                         (:children ((:node-type . :vector)
+                                     (:position . 5)
+                                     (:children ((:node-type . :number)
+                                                 (:position . 6)
+                                                 (:form . "123")
+                                                 (:value . 123))))
+                                    ((:node-type . :map)
+                                     (:position . 2)
+                                     (:children)))))))))
```
Comment (volrath) [inline test/parseclj-test.el]: shouldn't the `:map` happen first?

Comment (plexus) [reply test/parseclj-test.el]: ah yes, good catch!

File: parseclj-lex.el [hunk 2]
```diff
@@ -387,6 +442,22 @@ See `parseclj-lex-token'."
         (right-char)
         (parseclj-lex-token :rbrace "}" pos))
 
+       ((equal char ?')
+        (right-char)
+        (parseclj-lex-token :quote "'" pos))
+
+       ((equal char ?`)
+        (right-char)
+        (parseclj-lex-token :backquote "'" pos))
```
Comment (volrath) [inline parseclj-lex.el]: Also, I think there's a type here: ```"`"``` instead of ```"'"```

**Comment by bbatsov** [comment]
Btw, do you plan to tag a release now that the parser is "feature-complete"? As you know I'd love to use it and parseedn in CIDER, but there has to be some release available for use with MELPA stable.

**Comment by plexus** [comment]
Sure, sounds good! I just checked the [backlog](https://github.com/clojure-emacs/parseclj/issues/8) and it seems syntax-wise all we're missing are namespaced maps.

What do you have in mind for it? It might still be too slow for a lot of use cases. Parsing clojure.core to an AST takes 2.24 seconds for me. It's a lot faster than the old edn.el, but still...

OTOH we haven't really done any effort yet for optimizing things, maybe with some benching and tweaking we could get a significant speedup.

**Comment by plexus** [comment]
Note also that clojure.core is 7700+ lines. On a 200 line file I was just working on it runs in 0.086 seconds, that might already be workable.

**Comment by bbatsov** [comment]
> What do you have in mind for it? It might still be too slow for a lot of use cases. Parsing clojure.core to an AST takes 2.24 seconds for me. It's a lot faster than the old edn.el, but still...

I'm mostly hoping to implement better support for locals using it and some simple file-local refactorings. But what we need more right now is parseedn, so the cider-nrepl API could be unburned by the various Elisp limitations (e.g. some params are rather odd, because Elisp doesn't have booleans and some return values are also odd because the optimal return is not something we can read in Elisp - that's why almost all ops return just nested lists of strings).

**Comment by plexus** [comment]
I added namespace maps, AFAICT we should be able to parse all clojure syntax now. I'll merge this and tag a 0.1 release.

**Pull request merged**